src/isl/isl_type_reference.rs

```diff
@@ -143,6 +83,37 @@ impl IslTypeRef {
         }
     }
 
+    /// Return TypeId for given built-in type name from type_store
+    fn get_type_id_from_built_in_type_name(
+        alias: &str,
+        type_store: &mut TypeStore,
+        pending_types: &mut PendingTypes,
+    ) -> IonSchemaResult<TypeId> {
+        let invalid_type_reference_error = unresolvable_schema_error(format!(
+            "Could not resolve type reference: {:?} does not exist",
+            alias
+        ));
+
+        return if let Some(type_id) = type_store.get_builtin_type_id(alias) {
+            // verify if the given alias is a Built-in type and if it is then return the type id from type_store
+            // All Built-in types are preloaded into the type_store
+            // Built-in types includes all ion types and derived types like any, lob, text, number, $int, $float, ...
+            Ok(type_id)
+        } else if let Some(type_id) = pending_types.get_type_id_by_name(alias, type_store) {
+            // verify if the AliasType actually exists in the type_store or throw an error
+            Ok(type_id)
+        } else if let Some(parent) = pending_types.get_parent() {
+            // if it is a self referencing type resolve it using parent information from type_store
+            if parent.0.eq(alias) {
+                Ok(parent.1)
+            } else {
+                invalid_type_reference_error
+            }
+        } else {
+            invalid_type_reference_error
```

Comment:
Eagerly defining `invalid_type_reference_error` means that we'll be allocating/formatting an error string on every call even if nothing went wrong. I suggest defining these inline with messages that help distinguish between the two outcomes.

```suggestion
              unresolvable_schema_error(format!(
                "Could not resolve type reference: expected parent type to be {:?}",
                alias
              ));
            }
        } else {
            unresolvable_schema_error(format!(
              "Could not resolve type reference: {:?} does not exist",
              alias
            ));
```

I'm not sure I understand the first ("parent") error case, so provide a more meaningful error than my example :)

Reply:
parent is used for checking if the `type_reference` is a self reference. So if the current reference/alias is not same as parent reference/alias then the current type reference does not exist.
Added this change.